crypto/ec/ec_lib.c

```diff
@@ -318,17 +318,25 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
     } else
         BN_zero(&group->cofactor);
 
-    /*
-     * Some groups have an order with
-     * factors of two, which makes the Montgomery setup fail.
-     * |group->mont_data| will be NULL in this case.
+    /*-
+     * Access to the `mont_data` field of an EC_GROUP struct should always be
+     * guarded by an EC_GROUP_VERSION(group) check to avoid OOB accesses, as the
+     * group might come from the FIPS module, which does not define the
+     * `mont_data` field inside the EC_GROUP structure.
```

Comment:
I'm still a bit uneasy about the fact that obviously `EC_GROUP_set_generator()` is entered in FIPS mode instead of `FIPS_ec_group_set_generator()`. But there is apparently nothing one can do about it, looking at the callstack. 

```
#0  ec_precompute_mont_data (group=0x5555557766c0) at ec_lib.c:1111
#1  0x00007ffff7a1fdd8 in EC_GROUP_set_generator (group=0x5555557766c0, generator=0x555555776420,
    order=0x555555777d30, cofactor=0x555555777d70) at ec_lib.c:327
#2  0x00007ffff7a2a7d2 in ec_asn1_parameters2group (params=0x555555773220) at ec_asn1.c:912
...
```
In view of that, checking for `EC_GROUP_VERSION()` seems to be the best solution.

Reply:
This could be an alternative (similarly to, e.g., `EC_GROUP_new_curve_GFp()`):

~~~diff
diff --git a/crypto/ec/ec_lcl.h b/crypto/ec/ec_lcl.h
index 969fd147ef..92d308399d 100644
--- a/crypto/ec/ec_lcl.h
+++ b/crypto/ec/ec_lcl.h
@@ -565,4 +565,6 @@ EC_GROUP *FIPS_ec_group_new_curve_gfp(const BIGNUM *p, const BIGNUM *a,
 EC_GROUP *FIPS_ec_group_new_curve_gf2m(const BIGNUM *p, const BIGNUM *a,
                                        const BIGNUM *b, BN_CTX *ctx);
 EC_GROUP *FIPS_ec_group_new_by_curve_name(int nid);
+int FIPS_ec_group_set_generator(EC_GROUP *group, const EC_POINT *generator,
+                                const BIGNUM *order, const BIGNUM *cofactor);
 #endif
diff --git a/crypto/ec/ec_lib.c b/crypto/ec/ec_lib.c
index 933745248d..aed726b2bd 100644
--- a/crypto/ec/ec_lib.c
+++ b/crypto/ec/ec_lib.c
@@ -68,6 +68,10 @@

 #include "ec_lcl.h"

+#ifdef OPENSSL_FIPS
+# include <openssl/fips.h>
+#endif
+
 const char EC_version[] = "EC" OPENSSL_VERSION_PTEXT;

 /* functions for EC_GROUP objects */
@@ -293,6 +297,10 @@ int EC_METHOD_get_field_type(const EC_METHOD *meth)
 int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
                            const BIGNUM *order, const BIGNUM *cofactor)
 {
+#ifdef OPENSSL_FIPS
+    if(FIPS_mode())
+        return FIPS_ec_group_set_generator(group, generator, order, cofactor);
+#endif
     if (generator == NULL) {
         ECerr(EC_F_EC_GROUP_SET_GENERATOR, ERR_R_PASSED_NULL_PARAMETER);
         return 0;
~~~

But I'm under the impression that, even when `FIPS_mode() == 1`, there may be `EC_GROUP` and `FIPS_EC_GROUP` references flying around, so it is better to have a condition on `EC_GROUP_VERSION(group)` rather than on `FIPS_mode()`.
This also seems consistent with the fact that, currently, the only `FIPS_ec_group_*` methods definitions directly accessible to libcrypto code are the ones for creating a *new* group, while the handling of *existing* groups is done from `libcrypto` code, using `EC_GROUP_VERSION()` to handle differences in the `EC_GROUP` struct definition.